### PR TITLE
Add Node API in config/v1 

### PR DESCRIPTION
**Pull request by harche**
This PR adds the Node API in config/v1 that holds cluster-wide information about node specific features. 

Node API was[ proposed in the cgroups v2 enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/machine-config/mco-cgroupsv2-support.md#api-extensions) and later further modified in [WorkerLatencyProfile enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/worker-latency-profile/worker-latency-profile.md#api-extensions) to suite those use cases. 

Signed-off-by: Harshal Patil <harpatil@redhat.com>
Signed-off-by: Ryan Phillips <rphillip@redhat.com>
Signed-off-by: Swarup Ghosh <swghosh@redhat.com>

### CODE REVIEW

**Comment by rphillips** [comment]
/lgtm

This is what we agreed to in the enhancement.

File: config/v1/0000_10_config-operator_01_node.crd.yaml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  annotations:
+    api-approved.openshift.io: https://github.com/openshift/api/pull/1107
+    include.release.openshift.io/ibm-cloud-managed: "true"
+    include.release.openshift.io/self-managed-high-availability: "true"
+    include.release.openshift.io/single-node-developer: "true"
+  name: nodes.config.openshift.io
+spec:
+  group: config.openshift.io
+  names:
+    kind: Node
```
Comment (harche) [inline config/v1/0000_10_config-operator_01_node.crd.yaml:13]: We might have to rename this to something other than `Node`, something like `ConfigNode`. Kind `Node` already exists from `corev1`.


/hold

Comment (harche) [reply config/v1/0000_10_config-operator_01_node.crd.yaml:13]: Maybe it's fine. `Node` defined in this lives in `github.com/openshift/api/config` while the upstream `Node` comes from `k8s.io/api/core/v1`.

Comment (harche) [reply config/v1/0000_10_config-operator_01_node.crd.yaml:13]: /hold cancel

**Comment by harche** [comment]
Can't rebase successfully. Blocked on https://github.com/openshift/api/issues/1114 and https://github.com/openshift/api/issues/1113

**Comment by harche** [comment]
> /hold till #1114 is addressed.

Removing the hold because the issue is not related to PR. It's reproducible even without the changes from this RP. 

/hold cancel

File: config/v1/0000_10_config-operator_01_node.crd.yaml (hunk 1 above)
Comment (soltysh) [reply config/v1/0000_10_config-operator_01_node.crd.yaml:13]: Yup, it's fine.

File: config/v1/types_node.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+	// status holds observed values.
+	// +optional
+	Status NodeStatus `json:"status"`
+}
+
+type NodeSpec struct {
+	// CgroupMode determines the cgroups version on the node
+	// +optional
+	CgroupMode CgroupMode `json:"cgroupMode,omitempty"`
```
Comment (soltysh) [inline config/v1/types_node.go:30]: Nit: `CGroupMode` would be more correct

Comment (soltysh) [reply config/v1/types_node.go:30]: Actually, if you're only talking about version, why not just `CGroupVersion`?

File: config/v1/types_node.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	WorkerLatencyProfileStatus WorkerLatencyProfileStatus `json:"workerLatencyProfileStatus,omitempty"`
+}
+
+type CgroupMode string
+
+const (
+	CgroupMode_Empty   = "" // Empty string will always use the Default
+	CgroupMode_v1      = "v1"
+	CgroupMode_v2      = "v2"
+	CgroupMode_Default = CgroupMode_v1
```
Comment (soltysh) [inline config/v1/types_node.go:50]: Don't use `_` in names, we rarely do that, instead use `CGroupV1`, `CGroupV2`. I'd suggest dropping `CgroupMode_Default` entirely, same with `CgroupMode_empty` both are expressed as empty/not set. No need to add these as constants. Just be explicit that when not set default v1 is used on the field itself.

File: config/v1/types_node.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	// conditions describes the state of the WorkerLatencyProfile and related components
+	// (Kubelet or Controller Manager or Kube API Server)
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +optional
+	Conditions []WorkerLatencyStatusCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+
+	// relatedObjects is a list of objects that are "interesting" or related to this WorkerLatencyProfile. e.g. KubeletConfig object used for updating Kubelet arguments
+	// +optional
+	RelatedObjects []ObjectReference `json:"relatedObjects,omitempty"`
```
Comment (soltysh) [inline config/v1/types_node.go:77]: Why this is needed here?

**Review by soltysh — CHANGES_REQUESTED** [review]

File: config/v1/types_node.go (hunk 2 above)
Comment (harche) [reply config/v1/types_node.go:30]: @rphillips for the comment on cgroups

File: config/v1/types_node.go (hunk 3 above)
Comment (harche) [reply config/v1/types_node.go:50]: @rphillips for the comment on cgroups

Comment (rphillips) [reply config/v1/types_node.go:50]: This naming is carried over from the enhancement... The naming should stay the same.

File: config/v1/types_node.go (hunk 4 above)
Comment (harche) [reply config/v1/types_node.go:77]: Taking MCO as an example, but this might be true with `Kube Controller Manager Operator` and `Kube API Server Operator` with their respective operations related to `WorkerLatencyProfile`, 

1. If the `WorkerLatencyProfile` is deleted, then MCO needs to also delete the associated `KubeletConfig` 
2. MCO controller can get evicted in the middle of `WorkerLatecyProfile` rollout, MCO needs to know which `KubeletConfig` was associated the given `WorkerLatencyProfile` in a persistent way. 

In other words, there has to be a correlation maintained between the `WorkerLatencyProfile` and `KubeletConfig` (in case of MCO) in a persistent way.

File: config/v1/types_node.go (hunk 2 above)
Comment (soltysh) [reply config/v1/types_node.go:30]: I double checked k8s code, and I've noticed I was looking at comments but vars are using that, sorry for the noise :wink:

File: config/v1/types_node.go (hunk 3 above)
Comment (soltysh) [reply config/v1/types_node.go:50]: Fine with the naming, but the part about empty vs default still applies, both mean the exact same thing so I'd only add v1 and v2.

File: config/v1/types_node.go (hunk 4 above)
Comment (soltysh) [reply config/v1/types_node.go:77]: Not quite sure what you mean with that association, every controller and operator (being also a controller) should read the current state of the system and update it according to what is represented in spec. For that you don't need that association, you should always work with the current state, no matter what happened before, no?

Comment (harche) [reply config/v1/types_node.go]: I removed `RelatedObjects`, thanks.

File: config/v1/types_node.go (hunk 3 above)
Comment (rphillips) [reply config/v1/types_node.go:50]: @deads2k  mentioned adding support for empty and default.

Comment (soltysh) [reply config/v1/types_node.go:50]: Ack

Comment (soltysh) [reply config/v1/types_node.go:50]: There are still 3 nits to get it merged:
1. `CgroupModeEmpty` instead of `CgroupMode_Empty` to follow all the other types we have, ie. don't use `_` in the name
2. you need to explicitly tell that `CgroupModeEmpty` is of `CgroupMode`  iow. `CgroupModeEmpty CgroupMode = ""` w/o it's a string and not the type you wanted it to be
3. rename `CgroupMode` to `CgroupModeType` to follow other types like this.

@rphillips I was told to let you know about it :smile:

Comment (rphillips) [reply config/v1/types_node.go:50]: Thanks... +1

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
